### DISCUSSION TITLE
Shelly 2pm gen3 supported?

### DISCUSSION THREAD

**Discussion by joba-1** [arendst/Tasmota#22526] [Q&A] [answered]
Anyone successfully used this device with tasmota? Which template?

I searched Blackadders site and here and found close matches, but not exactly this one (2 relays with power moni, small)

According to the following link it uses an ESP-Shelly-C38F with 8MB flash.

https://kb.shelly.cloud/knowledge-base/shelly-2pm-gen3

**Comment by arendst** [comment]
Search the source code for the template. It works fine since months with tasmota.

Oops. No sorry. Haven't tested it yet.

**Comment by arendst** [comment]
Looking at the shelly specs I don't see why Tasmota this doesn't support already. It just needs someone to find the correct GPIO's and power monitoring hardware.

BTW Shelly is gone bunkers with regards to pricing; this device sells for 32,55 excl shipping. Their Dali Dimmer, in October 38,- now is 54,- and out of stock.

**Comment by joba-1** [comment]
I think I bite the bullet this time - free market at work. Couldn‘t find another device that small and every mm counts.

**Comment by arendst** [comment]
Yesterday I ordered latest gen3 stuff from proshop (demark). It'll take more time to receive but the prices are a lot lower than shellies.

**Comment by Jason2866** [comment]
Currently some Shelly devices are on Sale on Amazon

**Comment by arendst** [comment, marked as answer]
Template as of 20241129:
```
{"NAME":"Shelly 2PM Gen3","GPIO":[9472,3458,576,225,4736,224,640,608,1,1,193,0,0,0,0,0,0,0,192,32,1,1],"FLAG":0,"BASE":1}
```
with `ADCGPIO4 10000,10000,4000`

NOTE: 20241129 Updated template and ADCGPIO4 command

**Comment by joba-1** [comment]
please, how did you flash it? 

Yesterday I tried -|Rx|Tx|3V3|-|Gnd|Gnd found for older version.
Switched Rx and Tx, also changed direction.

Used an usb adapter that can usually deliver enough power for ESP32 dev boards.

There was short flashing of rx/tx leds of the adapter on connection, but no response after that. Also no AP coming up. 

Since testing time was up, for now it operates with the original firmware and works fine.

**Reply by Noschvie** [reply]
Did you connect GPIO0 to GND? Or try to use mgos32-to-tasmota32.

**Reply by joba-1** [reply]
if the pin assignment I used is correct, yes.
I‘ll try more, probably next week

**Comment by arendst** [comment]
> please, how did you flash it?

I use this https://www.tindie.com/products/voltlog/voltlink-cp2102n-usb-serial-adapter-programmer/ for flashing all shellies. It has an optional cable with connector fitting the tiny shelly connector.

**Comment by joba-1** [comment]
web flashing with above pins worked. I used unofficial C3 2MB which seems the closest precompiled image. Thanks! 

I applied (and activated) the template and adc command and now can toggle the relays and get a reasonable temperature.
But I get UV values for a VEML6070 and no energy values?

**Comment by joba-1** [comment]
ok, unofficial C3 2MB was the wrong image. Found the http://ota.tasmota.com/tasmota32/tasmota32c3.factory.bin.
Flashed with `esptool.py --chip auto --port /dev/ttyUSB0 --baud 921600 --before default_reset erase_flash` and `esptool.py --chip auto --port /dev/ttyUSB0 --baud 2000000 --before default_reset write_flash 0 ~/Downloads/tasmota32c3.factory.bin`
Now it works, except it shows 3164 volts. Calibrating... :)

**Comment by joba-1** [comment]
To round this up: [calibration](https://tasmota.github.io/docs/Power-Monitoring-Calibration/) worked - never seen an original value beeing so off, but hey...
For flashing I used the pins as [described here](https://kb.shelly.cloud/knowledge-base/shelly-2pm-gen3#Deviceelectricalinterfaces-Add-oninterface).
Of those pins I only connected GND and GPIO0 to GND, 3.3_ESP to 3.3V and U0RXD to TX, U0TXD to RX of my usb to serial adapter.
The reset pin is already pulled high, connect to gnd to reset. The DBG pin is not needed. 
For WLAN and tasmota configuration I had to **disconnect the pins for safety** and power the Shelly with 230V.

**Comment by elnappo** [comment]
Just for reference if someone is interested in pictures of the PCB with the ESP32-Shelly-C38F in v0.1.3 (B2445):
![Shelly 2PM Gen3 top](https://github.com/user-attachments/assets/f459095f-d4d0-4233-825b-a09a929b5e70)
![Shelly 2PM Gen3 bottom](https://github.com/user-attachments/assets/eb349268-a2b4-48e0-95eb-0200e5c8ef4d)
![Shelly 2PM Gen3 esp32](https://github.com/user-attachments/assets/cc2decfe-d930-494e-b298-79020aa3b884)

**Comment by Jason2866** [comment]
There is a header. Probably the usual Shelly setup. Wired flashing is easy.

**Reply by elnappo** [reply]
Yes it is, the pins are also reachable with the case on. Just don't forget to connect GPIO0 to GND while flashing.

**Reply by Noschvie** [reply]
Do you know if the Shelly Plus add-on can be used with it? Thanks

**Reply by elnappo** [reply]
No, I don't know that.

**Reply by markirb** [reply]
It can. Here is a Table that might help you that I have done for Shelly-HomeKit project using the addon on Gen3.
It has been used to run addons on 1G3 and 2PMG3

ESP Function | ESP32C3 GPIO# | ESP32 GPIO# | Used by addon as
-- | -- | -- | --
UART RX | 20 | 3 | ADDON_ADC_GPIO
UART TX | 21 | 1 | ADDON_IN_GPIO
Bootsel Input | 9 | 0 | ADDON_OUT_GPIO
Misc | 6 (1G3), 1 (2PMG3), others TBD | 19 | ADDON_DIG_INPUT

**Reply by Noschvie** [reply]
Is a software change in Tasmota required for this, or should it work the same way as with a Shelly Plus 1? Does anyone know?

**Comment by Haschi0** [comment]
I flashed firmware version 15.0.1.5 (31cc26e-tasmota32_2M) onto the Shelly 2PM Gen3, then applied the template as specified above. 
Now, the two relays are in the main menu, and I can control them. The temperature (ADC1) is also displayed.

However, there is no energy monitoring information. In the template, where I can see the GPIO ordering, there is no information for GPIO 0 and GPIO 1; they are completely empty. All the other GPIOs are set to something.
I know that the Shelly 2PM Gen3 has an ADE7953 chip for energy measurement but can't find it on the GPIO selection box. 

Does anyone know why the energy values are not shown in the main menu?

Any help will be welcome.

**Reply by Haschi0** [reply]
Hmmm, no one can help me with this.?

I might as well throw a rock at it and shoot it to the moon. I wanted to use it to measure my blind timings. However, to do that, I need to measure the power because the automatic system requires it to detect the endpoints.

**Comment by HH-HdG** [comment]
Hi there,

after some testing I found the following combination working for OneWire and both inputs with Shelly 2PM Gen3 and the Add-on:
| GPIO | function on the Add-On |
| --- | --- |
| GPIO1 | digital in |
| GPIO8 | ??? |
| GPIO9 | DS18x20_o |
| GPIO20 | digital in with a PWM duty cycle proportional with analog voltage on the analog in pin |
| GPIO21 | DS18x20 |

@arendst From my understanding xnrg_07_ade7953.ino uses "ADE7953 IRQ" pin assignment just to decide between different channel setups. Is it save to change the template to use red-flagged GPIO11 for it?

I used

`{"NAME":"Shelly 2PM Gen3","GPIO":[9472,1,576,225,4736,224,640,608,1,1,193,3458,0,0,0,0,0,0,192,32,1,1],"FLAG":0,"BASE":1,"CMND":"AdcGpio4 10000,10000,4000"}`

instead of 

`{"NAME":"Shelly 2PM Gen3","GPIO":[9472,3458,576,225,4736,224,640,608,1,1,193,0,0,0,0,0,0,0,192,32,1,1],"FLAG":0,"BASE":1,"CMND":"AdcGpio4 10000,10000,4000"}`

to get GPIO1 free for using as input on the Add-on.

**Comment by hopfi2k** [comment]
I still haven't managed to solve the wrong voltage display. Instead of ~230V I get 3140V...using the official template ([from here](https://templates.blakadder.com/shelly_plus_2PM.html), with PCB version v0.1.9). Anyone got that sorted?

Furthermore, I'm struggling to get S1 und S2 handled. It seems that they are not covered in the template?

Edit: and exposing S1/S2 to Matter as (input)switch would be the "cherry on the cake"

**Reply by sfromis** [reply]
That "official" template is not for for **gen3**, but instead for a different older device from Shelly. In this thread, arendst posted a template for "Shelly 2PM gen3", which should be the starting point, assuming that you have this *exact* device, but you did not post specific details.

**Reply by hopfi2k** [reply]
Thanks for helping out! I do have a Shelly 2PM Gen3 (PCB rev. 0.1.9) and already tried the template [mentioned earlier in this thread](https://github.com/arendst/Tasmota/discussions/22526#discussioncomment-11407244) that doesn't work at all (unfortunately). The one posted by @arendst with `ADCGPIO4` doesn't work either. I even tried the template @HH-HdG wrote about earlier, but with no success either.
Only the one posted on [blakadder](https://templates.blakadder.com/shelly_plus_2PM.html) gives me some functionality (with wrong voltage and no access to the input switches S1 and S2):
<img width="396" height="874" alt="Bildschirmfoto 2025-12-07 um 17 03 35" src="https://github.com/user-attachments/assets/e077606f-630e-4ad5-95ff-91e93fcdea61" />

I'd appreciate any help to get things sorted.

**Reply by hopfi2k** [reply]
well, entering `voltageset 230` in the console fixed the wrong voltage display...
